control and planning scripts/planner.py

```diff
@@ -0,0 +1,204 @@
+bot dimensions 34 cm x 45 cm
+
+TODO::::  Code optimisation
+'''
+########################################### section 1 imports
+
+import math
+import rospy
+import sys
+from nav_msgs.msg import Path
+from geometry_msgs.msg import PoseStamped
+from std_msgs.msg import Bool
+from geometry_msgs.msg import Pose
+from sensor_msgs.msg import Image
+from opencv_apps.msg import Contour
+from opencv_apps.msg import Point2D
+from nav_msgs.msg import Odometry
+
+############################################ section 2 defaults and global vars
+
+lookahead_distance = 1.5 
+swath_check_radius = 0.5
+r_sq = swath_check_radius**2
+d = sliding_dist = 0.15
+Emergency_stop = Bool()
+Emergency_stop = False
+
+waypoint_list_cropped = [] 
+slider_list = [0]*int(lookahead_distance/0.15)
+lookup = [0]*int(lookahead_distance/0.15)
+
+############################################ section 3 helper functions
+
+def collision_checker(xs, ys):
+	global lookup
+	global slider_list
+	global waypoint_list_cropped
+	#circle-point based checking
+	for i in range(len(waypoint_list_cropped)):
+		if not lookup[i]:
+			x_path = waypoint_list_cropped[i].x
+			y_path = waypoint_list_cropped[i].y
+			for j in range(len(xs)):
+				x_contour = xs[j]
+				y_contour = ys[j]
+				dist_sq = (x_contour - x_path)**2 + (y_contour - y_path)**2
+				if (r_sq >= dist_sq):
+					#sliding
+					print('in here collision detected so sliding')
+					all_set = slider(i, x_contour, y_contour, 1, xs, ys)
+					if not all_set:
+						Emergency_stop = True
+						emergency.publish(Emergency_stop)
+						return
+
+			lookup[i] = 1
+	for i in range(len(waypoint_list_cropped)-1):
+		if not lookup[i]:  #debug
+			all_set = False
+			print("oops! no path found")
+			return
+
+		else:
+			all_set = True
+	if all_set:
+		print("congratulations!!!! path found")
+		waypoints_out = Path()
+		waypoints_out.poses = []
+		for i in range(len(waypoint_list_cropped)-1):
+			pose = Pose()
+			pose.position.x = waypoint_list_cropped[i].x
+			pose.position.y = waypoint_list_cropped[i].y
+			waypoints_out.poses.append(pose)
+		pub.publish(waypoints_out)
+	return
+	#once a waypoint is slided then new path is generated between itself and next global wp so we need to check 
+	#for collision on that path by arbitration and again checking new path if its feasible. 
+	#so the checking process dhould terminate and then restart with new pathafter sliding.
+	
+def slider( i, x_contour, y_contour, check_bool, xs, ys ):
+	global lookup
+	global slider_list
+	global waypoint_list_cropped
+	if slider_list[i]:
+		if i > 1 and i < len(waypoint_list_cropped)-2:
+			b1 = slider(i-1, x_contour, y_contour, 0, xs, ys )
+			if b1: 
+				b2 = slider(i+1, x_contour, y_contour, 0,xs, ys)
+				if b2:
+					slider_list[i] = False
+					b3 = slider(i, x_contour, y_contour, 1, xs, ys )
+			if not b1 & b2 & b3:
+				return False
+		else:
+			print('path not possible better stop here')
+			return False
+	if not slider_list[i]:
```

Comment:
just else is fine? if you put another if condition, it will go through even if first if is True. Not very significant, but should be optimized.

Reply:
I'll check this once